compiler/p/codegen/ControlFlowEvaluator.cpp

```diff
@@ -1693,864 +2355,1027 @@ TR::Register *OMR::Power::TreeEvaluator::ifacmpneEvaluator(TR::Node *node, TR::C
+      // other, then check the sign of the result.
+      TR::Register *src1Reg = evaluateAndExtend(firstChild, compareInfo.isUnsigned, true, cg);
 
-// also handles bcmpgt, scmpgt
-TR::Register *OMR::Power::TreeEvaluator::icmpgtEvaluator(TR::Node *node, TR::CodeGenerator *cg)
-   {
-   TR::Node     *firstChild   = node->getFirstChild();
-   TR::Node     *secondChild  = node->getSecondChild();
-   TR::Register *src1Reg      = cg->evaluate(firstChild);
-   TR::Register *trgReg    = cg->allocateRegister();
-   if (secondChild->getOpCode().isLoadConst() &&
-       secondChild->getRegister() == NULL)
-      {
-      TR::Register    *temp1Reg = cg->allocateRegister();
-      int64_t value = secondChild->get64bitIntegralValue();
-      if (value == -1)
+      bool flipOrder = (compareInfo.cond == CompareCondition::gt || compareInfo.cond == CompareCondition::le);
+      bool flipResult = (compareInfo.cond == CompareCondition::le || compareInfo.cond == CompareCondition::ge);
+      bool isConst = secondChild->getOpCode().isLoadConst();
+
+      if (isConst && compareInfo.isUnsigned && secondChild->get64bitIntegralValue() < 0)
+         isConst = false;
+
+      if (flipOrder && isConst && is16BitSignedImmediate(secondChild->get64bitIntegralValue()))
+         {
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::subfic, node, trgReg, src1Reg, secondChild->get64bitIntegralValue());
+         }
+      else if (!flipOrder && isConst && is16BitSignedImmediate(-secondChild->get64bitIntegralValue()))
          {
-         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::srawi, node, temp1Reg, src1Reg, 31);
-         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addic, node, trgReg, temp1Reg, 1);
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, trgReg, src1Reg, -secondChild->get64bitIntegralValue());
          }
       else
          {
-         TR::Register *temp2Reg = addConstantToInteger(node, src1Reg, -(value+1), cg);
-         if (value >= 0)
-            {
-            generateTrg1Src2Instruction(cg, TR::InstOpCode::nor, node, temp1Reg, temp2Reg, src1Reg);
-            }
-         else // if (value < -1)
-            {
-            generateTrg1Src2Instruction(cg, TR::InstOpCode::nand, node, temp1Reg, temp2Reg, src1Reg);
-            }
-         cg->stopUsingRegister(temp2Reg);
-         generateShiftRightLogicalImmediate(cg, node, trgReg, temp1Reg, 31);
+         TR::Register *src2Reg = evaluateAndExtend(secondChild, compareInfo.isUnsigned, true, cg);
+
+         if (flipOrder)
+            generateTrg1Src2Instruction(cg, TR::InstOpCode::subf, node, trgReg, src1Reg, src2Reg);
+         else
+            generateTrg1Src2Instruction(cg, TR::InstOpCode::subf, node, trgReg, src2Reg, src1Reg);
+
+         stopUsingExtendedRegister(src2Reg, secondChild, cg);
          }
-      cg->stopUsingRegister(temp1Reg);
+
+      if (cg->comp()->target().is64Bit())
```

Comment:
This is a little bit tricky.  It is worth of a detailed comment.
Here, you treat is64Bit() as being the width of the register (actually not true). However, if you handle isSubRegisterType() correspondingly conservatively, this still can work out ok.

There is LPM (Live Partition Migration) implication too (on AIX, the process context for 32bit processes is really only 32bit, upon signal handling which includes LPM Dynamic-Reconfiguration events).

Reply:
I think this should be sufficiently explained by the changes I made to the Doxygen comment on `isTypeSubRegister`, as that makes it clear that we're treating GPRs as being of the same size as an address.